server/controllers/doppler.controller.js

```diff
@@ -0,0 +1,27 @@
+class DopplerController {
+    constructor(redisClientFactory) {
+      this.redisClientFactory = redisClientFactory;
+    }
+
+    async getShops(request, response) {
+        const {
+            session: { dopplerApiKey }
+        } = request;
+
+        const redis = this.redisClientFactory.createClient();
+        const shops = (await redis.getShopsAsync(dopplerApiKey, false))
+        .map(async shopName => {
+            let shop = await redis.getShopAsync(shopName);
+            return {
+                shop: shopName,
+                accessToken: shop.accessToken,
+                connectedOn: shop.connectedOn
+            };
+        });
+        
+        await redis.quitAsync();  
+        response.json(await Promise.all(shops));

```

Comment:
Oh, I understand now. It shouts all request "simultaneously" and then it waits for all of them. I am not sure if it is fine. What if there are a lot of shops and the remote server starts to deny requests?

Reply:
Interesting thought indeed. 
First, I think we should change the order of the lines 22 & 23. We should execute all the Redis requests and then close the connection.

> What if there are a lot of shops and the remote server starts to deny requests?

You mean the Redis server? If so, I think we should catch the error and return a null or something. The  function in the line 12 would become something like:

```javascript
const shops = (await redis.getShopsAsync(dopplerApiKey, false))
        .map(async shopName => {
            try {
              let shop = await redis.getShopAsync(shopName);
              return {
                  shop: shopName,
                  accessToken: shop.accessToken,
                  connectedOn: shop.connectedOn
              };
             } catch ( return null;);
        });
```

for the failed ones we should have nulls. 
What do you think?

Reply:
Oh, you are right about lines 22 and 23! there is a potential race condition there, right? I did not see that.

Regarding parallel requests, I have continued the discussion in [previous a thread](https://github.com/MakingSense/doppler-for-shopify/pull/17#discussion_r230349877).

And regarding exceptions, I was not talking about how to deal with them (In general, I prefer to ignore them over caught them silently like in your example). I was talking about exceptions generated because of the parallel requests approach, it is difficult for me to explain it, let's talk 3 minutes if I am not being clear.